projects/bootloader/test/test_ping.c

```diff
@@ -0,0 +1,182 @@
+  .error_event = DATAGRAM_EVENT_ERROR,
+  .error_cb = NULL,
+};
+
+#define TEST_DATA_LEN 32
+
+static uint8_t s_tx_data[TEST_DATA_LEN] = { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h',
+                                            'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p' };
+
+static uint8_t s_rx_data[TEST_DATA_LEN] = { 0 };
+static uint16_t s_rx_data_len;
+
+static CanMessage s_tx_buffer[TEST_TX_FIFO_SIZE];
+static Fifo s_tx_fifo;
+static bool s_tx_cmpl;
+static CanDatagramExitCb s_tx_cmpl_cb;
+
+static StatusCode prv_tx_to_fifo(uint8_t *data, size_t len, bool is_start_message) {
+  if (len > 8) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+  CanMessage message = {
+    .source_id = SYSTEM_CAN_DEVICE_BOOTLOADER,
+    .msg_id = TEST_CLIENT_SCRIPT_ID,
+    .type = is_start_message ? CAN_MSG_TYPE_ACK : CAN_MSG_TYPE_DATA,
+    .dlc = len,
+  };
+  // copy the message data over to the message to be transmitted
+  memcpy(message.data_u8, data, len);
+  fifo_push(&s_tx_fifo, &message);
+  return STATUS_CODE_OK;
+}
+
+static void prv_tx_from_fifo(void) {
+  CanMessage message;
+  fifo_pop(&s_tx_fifo, &message);
+  can_transmit(&message, NULL);
+}
+
+static void prv_mock_tx_compl(void) {
+  s_tx_cmpl = true;
+}
+
+static void prv_send_tx_to_fifo(CanDatagramTxConfig tx_config) {
+  fifo_init(&s_tx_fifo, s_tx_buffer);
+  s_tx_cmpl = false;
+
+  s_tx_cmpl_cb = tx_config.tx_cmpl_cb;
+  tx_config.tx_cmpl_cb = prv_mock_tx_compl;
+  tx_config.tx_cb = prv_tx_to_fifo;
+  can_datagram_start_tx(&tx_config);
+
+  Event e = { 0 };
+  while (!s_tx_cmpl) {
+    MS_TEST_HELPER_AWAIT_EVENT(e);
+    can_datagram_process_event(&e);
+    can_process_event(&e);
+  }  // puts all messages that are "sent" into the fifo
+  s_tx_cmpl_cb();
+}
+
+static CanMessage s_rx_buffer[TEST_TX_FIFO_SIZE];
+static Fifo s_rx_fifo;
+
+static StatusCode prv_rx_to_fifo(const CanMessage *msg, void *context, CanAckStatus *ack_reply) {
+  bool *callback = context;
+  *callback = true;
+
+  LOG_DEBUG("rxed: %i -> %i\n", msg->msg_id, msg->type);
+  if (msg->type == CAN_MSG_TYPE_ACK) {
+    fifo_init(&s_rx_fifo, s_rx_buffer);
+    fifo_push(&s_rx_fifo, msg);
+  } else {
+    fifo_push(&s_rx_fifo, msg);
+  }
+
+  return STATUS_CODE_OK;
+}
+
+void setup_test(void) {
+  event_queue_init();
+  interrupt_init();
+  soft_timer_init();
+  crc32_init();
+
+  bootloader_can_init(&s_test_can_storage, &s_test_can_settings, s_board_id);
+  can_datagram_init(&s_test_datagram_settings);
+  dispatcher_init(s_board_id);
+}
+
+void teardown_test(void) {}
+
+void test_ping(void) {
+  TEST_ASSERT_OK(ping_init(s_board_id));
+
+  CanDatagramTxConfig tx_config = {
+    .dgram_type = TEST_DATA_GRAM_ID,
+    .destination_nodes_len = 1,
+    .destination_nodes = &s_board_id,
+    .data_len = 0,
+    .data = NULL,
+    .tx_cmpl_cb = tx_cmpl_cb,
+  };
+  prv_send_tx_to_fifo(tx_config);
+
+  // this rx handler not working...
+  TEST_ASSERT_OK(can_register_rx_handler(s_board_id, prv_rx_to_fifo, NULL));
```

Comment:
Ah, the issue is that we can't actually do `can_register_rx_handler` with datagram messages. Because datagram messages have a particular arbitration ID format that isn't the same as the one used by our regular CAN system (see https://uwmidsun.atlassian.net/l/c/76h1GfbH, "CAN ID allocation" section), we actually added a special case in the CAN library (can_fsm.c, lines 58-62) which pass bootloader messages through the `bootloader_can` module rather than the regular CAN RX handler system.

Currently there can only be one registered bootloader CAN message handler, which is unfortunate. Since you're already changing bootloader_can, you could switch it to allow multiple RX handlers to be registered, and then you could use `bootloader_can_register_handler` here instead.

(On a somewhat unrelated note, it's great you've got all this infrastructure for testing with datagrams! Since it's repeated in both test_ping.c and test_dispatcher.c, it would be great to factor it out into a separate ms-common library that supports testing with datagrams. We have a few test-only libraries in ms-common, like `test_helpers.h` and `ms_test_helpers.h`, so you could have a `can_datagram_test_helpers` library. But that's for another ticket, let's get this merged first before doing that.)

Reply:
bootloader can actually checks if the sender is the client_script_board (msg_id 0) so it won't work for listening to controller to client script datagrams. I've added a separate callback for debuging that doesnt' check for client scipt id. And a separate register debug hander.